debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresReadOnlyIncrementalSnapshotChangeEventSource.java

```diff
@@ -0,0 +1,219 @@
+    }
+
+    private void readUntilNewTransactionChange(P partition, OffsetContext offsetContext) throws InterruptedException {
+
+        Long eventTxId = offsetContext.getSourceInfo().getInt64(SourceInfo.TXID_KEY);
+
+        LOGGER.debug("Event txId {}, snapshot is running {}, reachedHighWatermark {}",
+                eventTxId, getContext().snapshotRunning(), getContext().isTransactionVisible(eventTxId));
+
+        LOGGER.trace("Incremental snapshot context {}", getContext());
+        if (getContext().snapshotRunning() && maxInProgressTransactionCommitted(eventTxId)) {
+            getContext().closeWindow();
+            sendWindowEvents(partition, offsetContext);
+            readChunk(partition, offsetContext);
+
+            return;
+        }
+
+        while (getContext().snapshotRunning() && getContext().isTransactionVisible(eventTxId)) {
+
+            LOGGER.debug("Finishing snapshot, snapshot is running {}, reachedHighWatermark {}", getContext().snapshotRunning(),
+                    getContext().isTransactionVisible(eventTxId));
+
+            getContext().closeWindow();
+            sendWindowEvents(partition, offsetContext);
+            readChunk(partition, offsetContext);
+            if (getContext().watermarksChanged()) {
+                LOGGER.trace("Watermarks changed");
+                return;
+            }
+
+            LOGGER.trace("Re read chunk finished, snapshot is running {}, reachedHighWatermark {}", getContext().snapshotRunning(),
+                    getContext().isTransactionVisible(eventTxId));
+        }
+    }
+
+    private void getCurrentSnapshot(Consumer<PgSnapshot> watermark) {
+
+        try {
+
+            PgSnapshot pgSnapshot = jdbcConnection.queryAndMap(CURRENT_SNAPSHOT, singleResultMapper(rs -> {
+
+                String currentSnapshot = rs.getString(1);
+                LOGGER.trace("Current snapshot {}", currentSnapshot);
+
+                return PgSnapshot.from(currentSnapshot);
+            }, "Unable to get current snapshot"));
+
+            watermark.accept(pgSnapshot);
+        }
+        catch (SQLException e) {
+            throw new DebeziumException(e);
+        }
+    }
+
+    private boolean maxInProgressTransactionCommitted(Long eventTxId) {
+
+        if (getContext().getHighWatermark() == null) {
+            return false;
+        }
+
+        return getContext().getHighWatermark().getXMax().equals(eventTxId);
+    }
+
+    private void forceNewTransactionId() {
+        try {
+            jdbcConnection.query(FORCE_NEW_TRANSACTION, rs -> {
+                if (rs.next()) {
+                    LOGGER.trace("Created new transaction ID {}", rs.getString(1));
+                }
+            });
+        }
+        catch (SQLException e) {
+            throw new DebeziumException(e);
+        }
+    }
+
+    private <T> JdbcConnection.ResultSetMapper<T> singleResultMapper(JdbcConnection.ResultSetExtractor<T> extractor, String error) throws SQLException {
```

Comment:
Why can't you use `io.debezium.jdbc.JdbcConnection.singleResultMapper(ResultSetExtractor<T>, String)`?